logstash-core/src/test/java/org/logstash/plugin/InputTest.java

```diff
@@ -0,0 +1,46 @@
+package org.logstash.plugin;
+
+import org.junit.Test;
+import org.logstash.Event;
+import org.logstash.TestUtil;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.function.Consumer;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+public class InputTest {
+    private Collection<Collection<Event>> batches = new LinkedList<>();
+    private int batchCount = TestUtil.random.nextInt(10) + 1;
+
+    @Test
+    public void testInput() {
+        Input i = new TestInput();
+        i.run(batches::add);
+        assertEquals(batchCount, batches.size());
+        for (Collection<Event> events : batches) {
+            assertFalse(events.isEmpty());
+        }
+    }
+
+    private class TestInput implements Input {
+        @Override
+        public void run(Consumer<Collection<Event>> consumer) {
+            for (int b = 0; b < batchCount; b++) {
+                int eventCountPerBatch = TestUtil.random.nextInt(100) + 1;
+                Collection<Event> events = new LinkedList<>();
+                for (int i = 0; i < eventCountPerBatch; i++) {
+                    Event e = new Event();
```

Comment:
Here I am concerned about the separation of the datasource from the event build+decorate functions, how an author provides the two functions and how the inputs internal processing pipeline is built from standard parts and author provided parts (annotations).

something similar to (apologies, knocked up quick like)
```
    public interface Input {
        void run(Consumer<Collection<SourceData>> consumer);

        default Callable<Void> toCallable(Consumer<Collection<SourceData>> consumer) {
            return Executors.callable(() -> run(consumer), null);
        }
    }

    interface SourceData {
        // Toy definition
        SourceData add(String source);
        SourceData set(String key, Object value);
        Object get(String key);
        String source();
        Map<String, Object> details();
        SourceData cloneWithNewSource(String src);
    }

    @FunctionalInterface
    public interface EventGenerator {
        Stream<Event> accept(Stream<SourceData> s);
    }

    @FunctionalInterface
    public interface SourceDataProcessor {
        Stream<SourceData> accept(Stream<SourceData> t);
    }

    private class TestRecordIntoEvent implements EventGenerator {
        @Override
        public Stream<Event> accept(Stream<SourceData> provenanceStream) {
            return provenanceStream.map(provenance -> {
                Event e = new Event();
                e.setField("message", provenance.source());
                e.setField("@provenance", provenance.details());
                e.setField("i", provenance.get("iterationCount"));
                e.setField("b", provenance.get("batchCount"));
                return e;
            });
        }
    }

    private class TestCommaBoundaryDetector implements SourceDataProcessor {
        @Override
        public Stream<SourceData> accept(Stream<SourceData> provenanceStream) {
            return provenanceStream.flatMap(provenance -> Arrays.stream(provenance.source().split(",")).map(
                    provenance::cloneWithNewSource
            ));
        }
    }

    private class Provenance implements SourceData {
        // Toy impl
        // need an inner map of provenance key & values
        // a source string
        @Override
        public SourceData add(String source) { return this; }
        @Override
        public SourceData set(String key, Object value) { return this; }
        @Override
        public Object get(String key) { return null; }
        @Override
        public String source() { return null; }
        @Override
        public Map<String, Object> details() { return new HashMap<>(); }
        @Override
        public SourceData cloneWithNewSource(String src) { return new Provenance().add(src); }
    }

    private class TestInput implements Input {
        @Override
        public void run(Consumer<Collection<SourceData>> consumer) {
            int batchCount = 100;
            for (int b = 0; b < batchCount; b++) {
                int eventCountPerBatch = 20;
                Collection<SourceData> rawEvents = new LinkedList<>();
                for (int i = 0; i < eventCountPerBatch; i++) {
                    SourceData p = new Provenance();
                    p.add("hello,world,how,about,that,then");
                    p.set("iterationCount", i);
                    p.set("batchCount", b);
                    // other provenance settings
                    rawEvents.add(p);
                }
                consumer.accept(Collections.unmodifiableCollection(rawEvents));
            }
        }
    }
```

Reply:
This proposal makes sense to me -- having inputs produce "source data" (raw data) attached to some kind of Provenance (where the data came from).

Then the input ships this to some pipeline to pass it to the queue:

* input (source data) -> mill (milled data) -> decoder (decoded data) -> assembler (Event object) -> queue
* input (source data) -> mill (milled data) -> queue -> decoder (decoded data) -> assembler (Event object)

^^ seem like two options we could pursue. The 'Assembler' would take the decoded data plus the provenance and assemble an Event that the pipeline would handle?

*alternately* We maybe drop the idea of a formal "Event"?

Reply:
There is the question of how a plugin author codes an assembler with special handling, because while they can add certain key values to the provenance, there may still be the need to author the assembler which needs to be discoverable and inserted at the correct point (pre/post queue).

Reply:
Also, I have not seen many cases of composeable functions that take a stream and return a stream. However, this StackOverflow Q+A advocates for it. https://stackoverflow.com/a/24679745/5349531

I struggled with the idea of simply providing the function for `map` but then the CommaBoundaryDetector, being a split function, produced additional elements with the original provenance cloned with new source. The only (hurried) way I could think of is to take and return a stream.

This means that the Queue `accept` method would need to be a terminal function.

Reply:
Perhaps the Regex `Pattern.splitAsStream` would be a better choice.

However we have to handle the case where a value in a stream can produce a stream of values - and we have `flatMap` for that. But the knowledge of whether to use map or flatMap is only known to the function itself.